sgl-kernel/csrc/moe/moe_fused_gate.cu

```diff
@@ -81,57 +82,107 @@ __device__ void moe_fused_gate_impl(
+  T global_max_val_second = static_cast<T>(-FLT_MAX);
+  int global_max_idx = -1;
+  int global_max_second_idx = -1;
+
 #pragma unroll
-  for (int ii = 0; ii < params.VPT; ++ii) {
-    row_chunk[ii] = vec_thread_read_ptr[0][ii];
-    bias_chunk[ii] = vec_bias_thread_read_ptr[0][ii];
-  }
+  for (int tile = 0; tile < (params.VPT + 31) / 32; ++tile) {
+    int tile_offset = tile * 32;
+    int tile_size = min(32, params.VPT - tile_offset);
+    if (tile_size <= 0) break;
+
+    // Read row_chunk and bias_chunk
+    #pragma unroll
+      for (int ii = 0; ii < tile_size; ++ii) {
+        int global_idx = tile_offset + ii;
+        row_chunk[ii] = vec_thread_read_ptr[0][global_idx];
+        bias_chunk[ii] = vec_bias_thread_read_ptr[0][global_idx];
+      }
 
-  __syncthreads();
+      __syncthreads();
 
-////////////////////// Sigmoid //////////////////////
-#pragma unroll
-  for (int ii = 0; ii < params.VPT; ++ii) {
-    row_chunk[ii] = static_cast<T>(1.0f / (1.0f + expf(-float(row_chunk[ii]))));
-  }
-  __syncthreads();
+    ////////////////////// Sigmoid //////////////////////
+    #pragma unroll
+      for (int ii = 0; ii < tile_size; ++ii) {
+        row_chunk[ii] = static_cast<T>(1.0f / (1.0f + expf(-float(row_chunk[ii]))));
+      }
 
-////////////////////// Add Bias //////////////////////
-#pragma unroll
-  for (int ii = 0; ii < params.VPT; ++ii) {
-    bias_chunk[ii] = row_chunk[ii] + bias_chunk[ii];
+      __syncthreads();
+
+    ////////////////////// Add Bias //////////////////////
+    #pragma unroll
+      for (int ii = 0; ii < tile_size; ++ii) {
+        bias_chunk[ii] = row_chunk[ii] + bias_chunk[ii];
+      }
+
+      __syncthreads();
+
+      // 
+      if (tile == 0) {
+        // The first tile directly initializes the global maximum value
+        for (int ii = 0; ii < tile_size; ++ii) {
+          T val = bias_chunk[ii];
+          int global_idx = tile_offset + ii;
+          
+          if (cmp_gt(val, global_max_val)) {
+            global_max_val_second = global_max_val;
+            global_max_second_idx = global_max_idx;
+            global_max_val = val;
+            global_max_idx = global_idx;
+          } else if (cmp_gt(val, global_max_val_second)) {
+            global_max_val_second = val;
+            global_max_second_idx = global_idx;
+          }
+        }
+      } else {
+        // Subsequent tile, compare with the existing maximum value
+        for (int ii = 0; ii < tile_size; ++ii) {
+          T val = bias_chunk[ii];
+          int global_idx = tile_offset + ii;
+          
+          if (cmp_gt(val, global_max_val)) {
+            global_max_val_second = global_max_val;
+            global_max_second_idx = global_max_idx;
+            global_max_val = val;
+            global_max_idx = global_idx;
+          } else if (cmp_gt(val, global_max_val_second)) {
+            global_max_val_second = val;
+            global_max_second_idx = global_idx;
+          }
```

Comment:
I cannot see the difference between these two block

Reply:
Agree, there are a lot of duplicated code which can be refactored.